augur/curate/format_dates.py

```diff
@@ -0,0 +1,237 @@
+
+from augur.errors import AugurError
+from augur.io.print import print_err
+from augur.types import DataErrorMethod
+
+
+def register_parser(parent_subparsers):
+    parser = parent_subparsers.add_parser("format-dates",
+        parents=[parent_subparsers.shared_parser],
+        help=__doc__)
+
+    required = parser.add_argument_group(title="REQUIRED")
+    required.add_argument("--date-fields", nargs="+",
+        help="List of date field names in the record that need to be standardized.")
+    required.add_argument("--expected-date-formats", nargs="+",
+        help="Expected date formats that are currently in the provided date fields, " +
+             "defined by standard format codes as listed at " +
+             "https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes. " +
+             "If a date string matches multiple formats, it will be parsed as the first format in the list.")
+
+    optional = parser.add_argument_group(title="OPTIONAL")
+    optional.add_argument("--failure-reporting",
+        type=DataErrorMethod,
+        choices=[ method for method in DataErrorMethod ],
+        default=DataErrorMethod.ERROR_FIRST,
+        help="How should failed date formatting be reported.")
+    optional.add_argument("--no-mask-failure", dest="mask_failure",
+        action="store_false",
+        help="Do not mask dates with 'XXXX-XX-XX' and return original date string if date formatting failed.")
+
+    return parser
+
+
+def directive_is_included(potential_directives, date_format):
+    """
+    Checks if any of the directives in *potential_directives* is included
+    in *date_format* string.
+
+    If an element within *potential_directives* is a tuple, then all directives
+    within the tuple must be included in *date_format*.
+
+    Parameters
+    ----------
+    potential_directives: set[tuple[str, ...]]
+        Set of potential directives to check
+    date_format: str
+        Date format string to check for directives
+
+    Returns
+    -------
+    bool:
+        Whether the provided *date_format* includes any of the *potential_directives*
+
+
+    >>> potential_directives = {('%y', '%b', '%d'), ('%y', '%B', '%d'), ('%y', '%m', '%d'),}
+    >>> directive_is_included(potential_directives, '%G-%V-%A')
+    False
+    >>> directive_is_included(potential_directives, '%y-%m')
+    False
+    >>> directive_is_included(potential_directives, '%%y-%m-%d')
+    False
+    >>> directive_is_included(potential_directives, '%y-%m-%d')
+    True
+    >>> directive_is_included(potential_directives, '%y-%m-%dT%H:%M:%SZ')
+    True
+    """
+    return any(
+        all(
+            # Exclude escaped directives (e.g. '%%Y' means literal '%Y' not a four digit year)
+            bool(re.search(f"(?<!%){re.escape(sub_directive)}", date_format))
+            for sub_directive in directive
+        )
+        for directive in potential_directives
+    )
+
+
+def format_date(date_string, expected_formats):
```

Comment:
This function is similar to the format-based parsing that happens in [`AmbiguousDate`](https://github.com/nextstrain/augur/blob/master/augur/dates/ambiguous_date.py), except it isn't trying to resolve ambiguity into a range and supports more formats. I wonder if some code can be shared between the two.

Reply:
I remember looking through the existing dates module and thought they were doing pretty different things so I didn't want to force them to share code.

Reply:
Will revisit this as future improvement (https://github.com/nextstrain/augur/issues/1254)